extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -0,0 +1,322 @@
+    this.documentElement_.addEventListener('keydown', event => {
+      // Collapse mega menu on ESC.
+      if (event.key == Keys.ESCAPE) {
+        if (this.collapse_()) {
+          event.preventDefault();
+        }
+      }
+    });
+  }
+
+  /**
+   * Find all expandable menu items under this mega menu and add appropriate
+   * classes, attributes and event listeners to its children.
+   */
+  registerMenuItems_() {
+    this.items_ = toArray(scopedQuerySelectorAll(this.element, 'nav > * > li'));
+    this.items_.forEach(item => {
+      // skip if the item has already been registered.
+      if (item.classList.contains('i-amphtml-mega-menu-item')) {
+        return;
+      }
+      const heading =
+        scopedQuerySelector(item, '> button') ||
+        scopedQuerySelector(item, '> [role=button]');
+      const content = scopedQuerySelector(item, '> [role=dialog]');
+      // do not register item if either its heading or content element is missing,
+      // or if heading has on tap action.
+      if (
+        !heading ||
+        !content ||
+        this.action_.hasAction(heading, 'tap', item)
+      ) {
+        return;
+      }
+      item.classList.add('i-amphtml-mega-menu-item');
+      this.itemCount_++;
+
+      content.classList.add('i-amphtml-mega-menu-content');
+      content.setAttribute('aria-modal', 'false');
+      let contentId = content.getAttribute('id');
+      if (!contentId) {
+        // For accessibility, we need to make sure that each item has a unique ID.
+        // If ID is absent, we use a random number to ensure uniqueness.
+        contentId = this.prefix_ + '_AMP_content_' + this.itemCount_;
+        content.setAttribute('id', contentId);
+      }
+      content.insertBefore(
+        this.createScreenReaderCloseButton(),
+        content.firstChild
+      );
+      // prevent click event listener on document from closing the menu
+      content.addEventListener('click', e => e.stopPropagation());
+
+      heading.classList.add('i-amphtml-mega-menu-heading');
+      // haspopup value not set to menu since content can contain more than links.
+      heading.setAttribute('aria-haspopup', 'dialog');
+      heading.setAttribute('aria-controls', contentId);
+      if (!heading.hasAttribute('tabindex')) {
+        heading.setAttribute('tabindex', 0);
+      }
+      heading.setAttribute('aria-expanded', 'false');
+      heading.addEventListener('click', e => this.handleHeadingClick_(e));
+      // prevent focus on mousedown so that there's no sudden shift in focus
+      // when the content container opens.
+      heading.addEventListener('mousedown', e => e.preventDefault());
+      heading.addEventListener('keydown', e => this.handleHeadingKeyDown_(e));
+    });
+  }
+
+  /**
+   * Handler for item heading's on-click event to expand/collapse its content.
+   * @param {!Event} event click event.
+   */
+  handleHeadingClick_(event) {
+    event.preventDefault();
```

Comment:
Doesn't have to be in this PR, but we should define and document behavior for elements inside headings that have tap actions registered.